samples/extensions/ray_queries/ray_queries.cpp

```diff
@@ -0,0 +1,708 @@
+		{
+			if (accel->buffer)
+			{
+				accel->buffer.reset();
+			}
+			if (accel->handle)
+			{
+				vkDestroyAccelerationStructureKHR(device->get_handle(), accel->handle, nullptr);
+			}
+		}
+	}
+}
+
+void RayQueries::request_gpu_features(vkb::PhysicalDevice &gpu)
+{
+	RequestFeature(gpu)
+	    .request(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_BUFFER_DEVICE_ADDRESS_FEATURES, &VkPhysicalDeviceBufferDeviceAddressFeatures::bufferDeviceAddress)
+	    .request(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ACCELERATION_STRUCTURE_FEATURES_KHR, &VkPhysicalDeviceAccelerationStructureFeaturesKHR::accelerationStructure)
+	    .request(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_RAY_QUERY_FEATURES_KHR, &VkPhysicalDeviceRayQueryFeaturesKHR::rayQuery)
+	    .request(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_RAY_TRACING_PIPELINE_FEATURES_KHR, &VkPhysicalDeviceRayTracingPipelineFeaturesKHR::rayTracingPipeline);
+}
+
+void RayQueries::render(float delta_time)
+{
+	if (!prepared)
+	{
+		return;
+	}
+
+	draw();
+
+	update_uniform_buffers();
+}
+
+void RayQueries::build_command_buffers()
+{
+	VkCommandBufferBeginInfo command_buffer_begin_info = vkb::initializers::command_buffer_begin_info();
+
+	VkClearValue clear_values[2];
+	clear_values[0].color        = default_clear_color;
+	clear_values[1].depthStencil = {1.0f, 0};
+
+	VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
+	render_pass_begin_info.renderPass               = render_pass;
+	render_pass_begin_info.renderArea.offset.x      = 0;
+	render_pass_begin_info.renderArea.offset.y      = 0;
+	render_pass_begin_info.renderArea.extent.width  = width;
+	render_pass_begin_info.renderArea.extent.height = height;
+	render_pass_begin_info.clearValueCount          = 2;
+	render_pass_begin_info.pClearValues             = clear_values;
+
+	for (size_t i = 0; i < draw_cmd_buffers.size(); ++i)
+	{
+		render_pass_begin_info.framebuffer = framebuffers[i];
+
+		VK_CHECK(vkBeginCommandBuffer(draw_cmd_buffers[i], &command_buffer_begin_info));
+
+		vkCmdBeginRenderPass(draw_cmd_buffers[i], &render_pass_begin_info, VK_SUBPASS_CONTENTS_INLINE);
+
+		VkViewport viewport = vkb::initializers::viewport((float) width, (float) height, 0.0f, 1.0f);
+		vkCmdSetViewport(draw_cmd_buffers[i], 0, 1, &viewport);
+
+		VkRect2D scissor = vkb::initializers::rect2D(static_cast<int32_t>(width), static_cast<int32_t>(height), 0, 0);
+		vkCmdSetScissor(draw_cmd_buffers[i], 0, 1, &scissor);
+
+		vkCmdBindPipeline(draw_cmd_buffers[i], VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline);
+		vkCmdBindDescriptorSets(draw_cmd_buffers[i], VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_layout, 0, 1, &descriptor_set, 0, nullptr);
+		VkDeviceSize offsets[1] = {0};
+		vkCmdBindVertexBuffers(draw_cmd_buffers[i], 0, 1, vertex_buffer->get(), offsets);
+		vkCmdBindIndexBuffer(draw_cmd_buffers[i], index_buffer->get_handle(), 0, VK_INDEX_TYPE_UINT32);
+		vkCmdDrawIndexed(draw_cmd_buffers[i], static_cast<uint32_t>(model.indices.size()) * 3, 1, 0, 0, 0);
```

Comment:
It seems that we are aggregating all geometry and dispatching a single draw call for the whole scene. Perhaps a more realistic use case could be to have a deferred rendering pipeline, to use the same draw calls to render both the scene and the shadows, since the ray queries can then be traced once per pixel during the lighting pass

Reply:
I'm less certain about this, it's a single model.  A good optimization for most use cases is to reduce the number of draw calls and this is emblematic of that.  However, at the same time it would make sense to use space partitioning given that the model is a full scene one can traverse.  The complexity rises to do this yet doesn't refine the actual sample's use case.
Maybe a good compromise would be to add a model to the scene that explicitly could be a good use case for another draw call?  What are your thoughts?

Reply:
I was more concerned about overdraw, since we might be unnecessarily running ray queries for occluded pixels. A deferred pass would limit it to one ray per pixel. I can see how this would add complexity to the sample though.

Reply:
Overdraw is a compelling reason.  I have to admit to being torn as I think simple samples are better than complex ones to communicate the core idea.  However, a reasoning for ray queries to exist in a common pipeline would be to be able to make use of running them only in a deferred fragment pass so they are limited in scope to the number of pixels actually to be rendered.
I'm leaning towards saying this is beyond the scope of a basic introduction to ray queries as a core tenet of the extension is they can be used from any shader at any point in either the traditional or ray tracing pipelines.  However, a performance sample would make a lot of sense to demonstrate the reasoning why overdraw would be a potential issue.  Maybe a future sample should address this.  Do you have any thoughts on that approach?

Reply:
> a reasoning for ray queries to exist in a common pipeline would be to be able to make use of running them only in a deferred fragment pass so they are limited in scope to the number of pixels actually to be rendered

Yes exactly, I was going for a simple deferred pipeline like in some of the performance samples, to show how ray queries can be part of it. But also happy to let this sample focus on the use of the extension and address more realistic use-cases in future samples